[V1][Spec Decoding] Use model_loader.get_model() to load models

markmc opened the pull request:
With:

```
$> vllm serve lmsys/vicuna-7b-v1.3 --speculative-config '{"method": "medusa", "model": "abhigoyal/vllm-medusa-vicuna-7b-v1.3", "num_speculative_tokens": 6}'
```

(A misconfiguration: num_spec_tokens > num_heads)

I noticed in V0 we get this fairly obscure error:

```
  File "vllm/vllm/model_executor/model_loader/default_loader.py", line 294, in load_model
    raise ValueError(
ValueError: Following weights were not initialized from checkpoint: {'blocks.5.layers.0.weight', 'lm_heads.5.weight'}
```

but in V1, we silently accept the config.

It turns out the difference is that in V0 we're using `ModelLoader.load_model()` but in V1 we're reimplementing a subset that logic in the Medusa and EAGLE proposers.
    
If we use the full model loading infrastructure, we can reduce the complexity of drafters, and automatically have support for e.g. quantized weights as per #18290.

@WoosukKwon @LiuXiaoxuanPKU @skylee-01

markmc committed:
[V1][Spec Decoding] Use model_loader.get_model() to load models

With:

```
$> vllm serve lmsys/vicuna-7b-v1.3 --speculative-config '{"method": "medusa", "model": "abhigoyal/vllm-medusa-vicuna-7b-v1.3", "num_speculative_tokens": 6}'
```

(A misconfiguration: num_spec_tokens > num_heads)

I noticed in V0 we get this fairly obscure error:

  File "/home/markmc/vllm-project/vllm/vllm/model_executor/model_loader/default_loader.py", line 294, in load_model
    raise ValueError(
ValueError: Following weights were not initialized from checkpoint: {'blocks.5.layers.0.weight', 'lm_heads.5.weight'}

but in V1, we silently accept the config.

It turns out the difference is that in V0 we're using
ModelLoader.load_model() but in V1 we're reimplementing
a subset that logic in the Medusa and EAGLE proposers.

If we use the full model loading infrastructure, we can
reduce the complexity of drafters, and automatically
have support for e.g. quantized weights as per #18290.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
[V1][Spec Decode] Fix v1/spec_decode/test_eagle.py

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc commented on tests/v1/spec_decode/test_eagle.py:
```diff
@@ -117,34 +117,13 @@ def test_prepare_inputs():
-def test_load_model(mock_set_config, mock_set_dtype, mock_get_loader,
-                    mock_registry, mock_get_layers, mock_get_pp_group, method,
+@mock.patch('vllm.v1.spec_decode.eagle.get_model')
+def test_load_model(mock_get_model, mock_get_layers, mock_get_pp_group, method,
```
This kind of unit test where everything is mocked, where we're verifying implementation and not behavior, and where the test is a brittle mirror of the implementation ... is not useful IMHO

wwl2755 replied:
Yes. I kind of agree :joy: Sorry for the bad design.

As you already unified the `get_model()` function, and `load_model` is basically calling `get_model` in the eagle, maybe the `test_load_model` could be now removed here.

WoosukKwon reviewed (approved):
LGTM! @markmc Could you please rebase the PR and re-run the CI?

WoosukKwon commented:
@markmc I've rebased it myself for faster merge. Let's see if it passes the CI.

WoosukKwon commented:
@markmc The CI error looks related to this PR. Please take a look?

markmc committed:
[V1][Spec Decode] Fix eagle3 parameter definition

The redundant .type(torch.LongTensor) moves the GPU tensor to CPU
causing:

```
    targets = base + self.draft_id_to_target_id
              ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~

RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!
```

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc commented:
Looks like the issue was in `llama_eage3.py`:

```
         self.draft_id_to_target_id = nn.Parameter(
-            torch.zeros((self.config.draft_vocab_size),
-                        dtype=torch.long).type(torch.LongTensor),
+            torch.zeros(self.config.draft_vocab_size, dtype=torch.long),
```

The redundant `.type(LongTensor)` was moving the GPU tensor to CPU

This was masked previously by the `model.to()`

```
-            self.model = draft_model_cls(
-                vllm_config=self.vllm_config,
-                start_layer_id=target_layer_num).to(target_device)
```
